File: packages/schema-domain/src/app-game-adapter-dispatch-result.ts
import { AppGameSchemaVersion } from './app-game-primitives';
import {
  AgentAppGameAdapterDispatchDecision,
  AgentAppGameAdapterDispatchOutcomeState,
  AgentAppGameAdapterDispatchPreflightState,
} from './app-game-adapter-dispatch-preflight';
import { type Infer, NonEmptyStringSchema, Schema, withParser } from './effect';

const DispatchResultCount = Schema.Number.pipe(Schema.nonNegative(), Schema.int());
const AgentEnforcementExecuteCommand = 'agent.enforcement.execute' as const;
const AgentEnforcementAuditReportedEvent = 'agent.enforcement.audit.reported' as const;
const AgentActivityAppGameAdapterDispatchResultReadModelGetCommand =
  'agent.activity.app-game.adapter-dispatch-result.read-model.get' as const;

export const AgentAppGameAdapterDispatchResultPayloadField = 'appGameAdapterDispatchResultReadModel' as const;
export const AgentAppGameAdapterDispatchExecuteResultPayloadField = 'appGameAdapterDispatchExecuteResult' as const;

export const AgentAppGameAdapterDispatchCommandResultState = {
  CommandAccepted: 'command-accepted',
  BlockedBeforeCommand: 'blocked-before-command',
  ManualRequired: 'manual-required',
  Unavailable: 'unavailable',
  Unsupported: 'unsupported',
  Degraded: 'degraded',
} as const;

export const AgentAppGameAdapterDispatchCommandResultDecision = {
  CommandAccepted: 'command-accepted',
  BlockedBeforeCommand: 'blocked-before-command',
} as const;

export const AgentAppGameAdapterDispatchExecutionAuditState = {
  ServiceLocalAuditRecorded: 'service-local-audit-recorded',
  BlockedBeforeExecutionAudit: 'blocked-before-execution-audit',
} as const;

export const AgentAppGameAdapterDispatchExecutionAuditDecision = {
  ServiceLocalAuditRecorded: 'service-local-audit-recorded',
  BlockedBeforeExecutionAudit: 'blocked-before-execution-audit',
} as const;

export const AgentAppGameAdapterDispatchAdapterExecutionState = {
  AdapterExecutionReported: 'adapter-execution-reported',
  AdapterExecutionEvidenceMissing: 'adapter-execution-evidence-missing',
  BlockedBeforeAdapterExecution: 'blocked-before-adapter-execution',
} as const;

export const AgentAppGameAdapterDispatchAdapterExecutionDecision = {
  AdapterExecutionReported: 'adapter-execution-reported',
  AdapterExecutionEvidenceMissing: 'adapter-execution-evidence-missing',
  BlockedBeforeAdapterExecution: 'blocked-before-adapter-execution',
} as const;

const AgentAppGameAdapterDispatchResultRowBaseSchema = Schema.Struct({
  schemaVersion: Schema.Literal(AppGameSchemaVersion),
  rowId: NonEmptyStringSchema,
  sourceDispatchPreflightRowId: NonEmptyStringSchema,
  sourceProofEntryId: NonEmptyStringSchema,
  platform: NonEmptyStringSchema,
  productMeanings: Schema.Array(Schema.Literal('native-app', 'native-game')),
  adapterCapability: NonEmptyStringSchema,
  dispatchPreflightState: Schema.Literal(
    AgentAppGameAdapterDispatchPreflightState.DispatchEligible,
    AgentAppGameAdapterDispatchPreflightState.BlockedBeforeDispatch,
    AgentAppGameAdapterDispatchPreflightState.ManualRequired,
    AgentAppGameAdapterDispatchPreflightState.Unavailable,
    AgentAppGameAdapterDispatchPreflightState.Unsupported,
    AgentAppGameAdapterDispatchPreflightState.Degraded
  ),
  dispatchDecision: Schema.Literal(
    AgentAppGameAdapterDispatchDecision.DispatchEligible,
    AgentAppGameAdapterDispatchDecision.BlockedBeforeDispatch
  ),
  dispatchIntentId: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchOutcomeState: Schema.Literal(
    AgentAppGameAdapterDispatchOutcomeState.DispatchReady,
    AgentAppGameAdapterDispatchOutcomeState.ManualRequired,
    AgentAppGameAdapterDispatchOutcomeState.Unavailable,
    AgentAppGameAdapterDispatchOutcomeState.Unsupported,
    AgentAppGameAdapterDispatchOutcomeState.Degraded,
    AgentAppGameAdapterDispatchOutcomeState.NotDispatched
  ),
  dispatchCommandResultState: Schema.Literal(
    AgentAppGameAdapterDispatchCommandResultState.CommandAccepted,
    AgentAppGameAdapterDispatchCommandResultState.BlockedBeforeCommand,
    AgentAppGameAdapterDispatchCommandResultState.ManualRequired,
    AgentAppGameAdapterDispatchCommandResultState.Unavailable,
    AgentAppGameAdapterDispatchCommandResultState.Unsupported,
    AgentAppGameAdapterDispatchCommandResultState.Degraded
  ),
  dispatchCommandResultDecision: Schema.Literal(
    AgentAppGameAdapterDispatchCommandResultDecision.CommandAccepted,
    AgentAppGameAdapterDispatchCommandResultDecision.BlockedBeforeCommand
  ),
  enforcementCommandName: Schema.Union(Schema.Literal(AgentEnforcementExecuteCommand), Schema.Null),
  enforcementEventName: Schema.Union(Schema.Literal(AgentEnforcementAuditReportedEvent), Schema.Null),
  enforcementActionMode: Schema.Union(Schema.Literal('terminate-process'), Schema.Null),
  dispatchCommandResultId: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchCommandAuditRefs: Schema.Array(NonEmptyStringSchema),
  dispatchCommandTimerRefs: Schema.Array(NonEmptyStringSchema),
  dispatchExecutionAuditState: Schema.Literal(
    AgentAppGameAdapterDispatchExecutionAuditState.ServiceLocalAuditRecorded,
    AgentAppGameAdapterDispatchExecutionAuditState.BlockedBeforeExecutionAudit
  ),
  dispatchExecutionAuditDecision: Schema.Literal(
    AgentAppGameAdapterDispatchExecutionAuditDecision.ServiceLocalAuditRecorded,
    AgentAppGameAdapterDispatchExecutionAuditDecision.BlockedBeforeExecutionAudit
  ),
  dispatchExecutionAuditId: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchExecutionAuditRefs: Schema.Array(NonEmptyStringSchema),
  dispatchAdapterExecutionState: Schema.Literal(
    AgentAppGameAdapterDispatchAdapterExecutionState.AdapterExecutionReported,
    AgentAppGameAdapterDispatchAdapterExecutionState.AdapterExecutionEvidenceMissing,
    AgentAppGameAdapterDispatchAdapterExecutionState.BlockedBeforeAdapterExecution
  ),
  dispatchAdapterExecutionDecision: Schema.Literal(
    AgentAppGameAdapterDispatchAdapterExecutionDecision.AdapterExecutionReported,
    AgentAppGameAdapterDispatchAdapterExecutionDecision.AdapterExecutionEvidenceMissing,
    AgentAppGameAdapterDispatchAdapterExecutionDecision.BlockedBeforeAdapterExecution
  ),
  dispatchAdapterExecutionResultId: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchAdapterExecutionStatus: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchAdapterExecutionAdapterResultCode: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchAdapterExecutionAuditEventId: Schema.Union(NonEmptyStringSchema, Schema.Null),
  dispatchAdapterExecutionRefs: Schema.Array(NonEmptyStringSchema),
  manualProofRequirements: Schema.Array(NonEmptyStringSchema),
  claimBoundary: NonEmptyStringSchema,
  fallbackBehavior: NonEmptyStringSchema,
  adapterDispatchCommandResultClaimed: Schema.Boolean,
  adapterDispatchExecutedClaimed: Schema.Boolean,
  serviceLocalExecutionAuditClaimed: Schema.Boolean,
  broadInstalledAppBlockingClaimed: Schema.Literal(false),
  childDeviceDeliveryClaimed: Schema.Literal(false),
  platformEnforcementClaimed: Schema.Literal(false),
  providerDeliveryClaimed: Schema.Literal(false),
  privateDiagnosticsClaimed: Schema.Literal(false),
  lastCheckedAt: NonEmptyStringSchema,
});

type AgentAppGameAdapterDispatchResultRowCandidate = Infer<typeof AgentAppGameAdapterDispatchResultRowBaseSchema>;

export const AgentAppGameAdapterDispatchResultRowSchema = withParser(
  AgentAppGameAdapterDispatchResultRowBaseSchema.pipe(
    Schema.filter(
      (row: AgentAppGameAdapterDispatchResultRowCandidate) =>
        dispatchResultRowIsHonest(row) ||
        'Expected only the scoped Windows owned-process time-limit row to claim an accepted dispatch ' +
          'command result without claiming adapter execution'
    )
  )
);

const AgentAppGameAdapterDispatchResultReadModelBaseSchema = Schema.Struct({
  schemaVersion: Schema.Literal(AppGameSchemaVersion),
  readModelId: NonEmptyStringSchema,
  generatedAt: NonEmptyStringSchema,
  sourceReadModelIds: Schema.Array(NonEmptyStringSchema),
  custodyLabel: NonEmptyStringSchema,
  capabilityStatus: NonEmptyStringSchema,
  returned: DispatchResultCount,
  commandAcceptedCount: DispatchResultCount,
  blockedBeforeCommandCount: DispatchResultCount,
  executionAuditRecordedCount: DispatchResultCount,
  blockedBeforeExecutionAuditCount: DispatchResultCount,
  adapterExecutionReportedCount: DispatchResultCount,
  adapterExecutionEvidenceMissingCount: DispatchResultCount,
  blockedBeforeAdapterExecutionCount: DispatchResultCount,
  adapterDispatchCommandResultClaimedCount: DispatchResultCount,
  serviceLocalExecutionAuditClaimedCount: DispatchResultCount,
  adapterDispatchExecutedClaimedCount: DispatchResultCount,
  broadInstalledAppBlockingClaimed: Schema.Literal(false),
  childDeviceDeliveryClaimed: Schema.Literal(false),
  platformEnforcementClaimed: Schema.Literal(false),
  providerDeliveryClaimed: Schema.Literal(false),
  privateDiagnosticsClaimed: Schema.Literal(false),
  rows: Schema.Array(AgentAppGameAdapterDispatchResultRowSchema),
});

type AgentAppGameAdapterDispatchResultReadModelCandidate = Infer<
  typeof AgentAppGameAdapterDispatchResultReadModelBaseSchema
>;

export const AgentAppGameAdapterDispatchResultReadModelSchema = withParser(
  AgentAppGameAdapterDispatchResultReadModelBaseSchema.pipe(
    Schema.filter(
      (readModel: AgentAppGameAdapterDispatchResultReadModelCandidate) =>
        dispatchResultReadModelIsConsistent(readModel) ||
        'Expected app/game adapter dispatch result counts and row ids to match the rows'
    )
  )
);

const AgentAppGameAdapterDispatchExecuteResultBaseSchema = Schema.Struct({
  schemaVersion: Schema.Literal(AppGameSchemaVersion),
  commandId: NonEmptyStringSchema,
  generatedAt: NonEmptyStringSchema,
  sourceReadModelId: Schema.Literal('app-game-adapter-dispatch-result'),
  sourceDispatchRowId: Schema.Literal('app-game-adapter-dispatch-result-windows-app-game-owned-process-time-limit'),
  sourceProofEntryId: Schema.Literal('windows-app-game-owned-process-time-limit'),
  executionCommandName: Schema.Literal(AgentEnforcementExecuteCommand),
  executionEventName: Schema.Literal(AgentEnforcementAuditReportedEvent),
  executionResultId: NonEmptyStringSchema,
  executionStatus: NonEmptyStringSchema,
  executionAdapterResultCode: NonEmptyStringSchema,
  executionAuditEventId: NonEmptyStringSchema,
  readbackCommandName: Schema.Literal(AgentActivityAppGameAdapterDispatchResultReadModelGetCommand),
  adapterDispatchExecutedClaimed: Schema.Boolean,
  broadInstalledAppBlockingClaimed: Schema.Literal(false),
  childDeviceDeliveryClaimed: Schema.Literal(false),
  platformEnforcementClaimed: Schema.Literal(false),
  providerDeliveryClaimed: Schema.Literal(false),
  privateDiagnosticsClaimed: Schema.Literal(false),
});

type AgentAppGameAdapterDispatchExecuteResultCandidate = Infer<
  typeof AgentAppGameAdapterDispatchExecuteResultBaseSchema
>;

export const AgentAppGameAdapterDispatchExecuteResultSchema = withParser(
  AgentAppGameAdapterDispatchExecuteResultBaseSchema.pipe(
    Schema.filter(
      (result: AgentAppGameAdapterDispatchExecuteResultCandidate) =>
        result.executionStatus === 'actually-enforced' ||
        !result.adapterDispatchExecutedClaimed ||
        'Expected scoped adapter dispatch execution to claim execution only when enforcement reports actually-enforced'
    )
  )
);

export type AgentAppGameAdapterDispatchResultRow = Infer<typeof AgentAppGameAdapterDispatchResultRowSchema>;
export type AgentAppGameAdapterDispatchResultReadModel = Infer<typeof AgentAppGameAdapterDispatchResultReadModelSchema>;
export type AgentAppGameAdapterDispatchExecuteResult = Infer<typeof AgentAppGameAdapterDispatchExecuteResultSchema>;

function dispatchResultRowIsHonest(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return row.dispatchCommandResultState === AgentAppGameAdapterDispatchCommandResultState.CommandAccepted
    ? acceptedDispatchResultRowIsHonest(row)
    : blockedDispatchResultRowIsHonest(row);
}

function acceptedExecutionEvidenceIsHonest(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return row.dispatchAdapterExecutionDecision ===
    AgentAppGameAdapterDispatchAdapterExecutionDecision.AdapterExecutionReported
    ? reportedExecutionEvidenceIsHonest(row)
    : missingExecutionEvidenceIsHonest(row);
}

function dispatchResultReadModelIsConsistent(readModel: AgentAppGameAdapterDispatchResultReadModelCandidate): boolean {
  const countExpectations = [
    {
      expected: readModel.commandAcceptedCount,
      actual: readModel.rows.filter(
        (row) => row.dispatchCommandResultDecision === AgentAppGameAdapterDispatchCommandResultDecision.CommandAccepted
      ).length,
    },
    {
      expected: readModel.blockedBeforeCommandCount,
      actual: readModel.rows.filter(
        (row) =>
          row.dispatchCommandResultDecision === AgentAppGameAdapterDispatchCommandResultDecision.BlockedBeforeCommand
      ).length,
    },
    {
      expected: readModel.executionAuditRecordedCount,
      actual: readModel.rows.filter(
        (row) =>
          row.dispatchExecutionAuditDecision ===
          AgentAppGameAdapterDispatchExecutionAuditDecision.ServiceLocalAuditRecorded
      ).length,
    },
    {
      expected: readModel.blockedBeforeExecutionAuditCount,
      actual: readModel.rows.filter(
        (row) =>
          row.dispatchExecutionAuditDecision ===
          AgentAppGameAdapterDispatchExecutionAuditDecision.BlockedBeforeExecutionAudit
      ).length,
    },
    {
      expected: readModel.adapterExecutionReportedCount,
      actual: readModel.rows.filter(
        (row) =>
          row.dispatchAdapterExecutionDecision ===
          AgentAppGameAdapterDispatchAdapterExecutionDecision.AdapterExecutionReported
      ).length,
    },
    {
      expected: readModel.adapterExecutionEvidenceMissingCount,
      actual: readModel.rows.filter(
        (row) =>
          row.dispatchAdapterExecutionDecision ===
          AgentAppGameAdapterDispatchAdapterExecutionDecision.AdapterExecutionEvidenceMissing
      ).length,
    },
    {
      expected: readModel.blockedBeforeAdapterExecutionCount,
      actual: readModel.rows.filter(
        (row) =>
          row.dispatchAdapterExecutionDecision ===
          AgentAppGameAdapterDispatchAdapterExecutionDecision.BlockedBeforeAdapterExecution
      ).length,
    },
    {
      expected: readModel.adapterDispatchCommandResultClaimedCount,
      actual: readModel.rows.filter((row) => row.adapterDispatchCommandResultClaimed).length,
    },
    {
      expected: readModel.serviceLocalExecutionAuditClaimedCount,
      actual: readModel.rows.filter((row) => row.serviceLocalExecutionAuditClaimed).length,
    },
    {
      expected: readModel.adapterDispatchExecutedClaimedCount,
      actual: readModel.rows.filter((row) => row.adapterDispatchExecutedClaimed).length,
    },
  ] as const;

  return (
    readModel.returned === readModel.rows.length &&
    countExpectations.every(({ expected, actual }) => expected === actual) &&
    new Set(readModel.rows.map((row) => row.rowId)).size === readModel.rows.length
  );
}

function acceptedDispatchResultRowIsHonest(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    acceptedDispatchResultMatchesBoundary(row) &&
    acceptedDispatchResultTracksCommand(row) &&
    acceptedDispatchResultTracksExecutionAudit(row) &&
    acceptedDispatchResultKeepsClaimsScoped(row)
  );
}

function blockedDispatchResultRowIsHonest(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    blockedDispatchResultTracksCommand(row) &&
    blockedDispatchResultTracksExecutionAudit(row) &&
    blockedDispatchResultTracksAdapterExecution(row) &&
    blockedDispatchResultKeepsClaimsScoped(row)
  );
}

function reportedExecutionEvidenceIsHonest(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.dispatchAdapterExecutionState === AgentAppGameAdapterDispatchAdapterExecutionState.AdapterExecutionReported &&
    row.dispatchAdapterExecutionResultId !== null &&
    row.dispatchAdapterExecutionStatus === 'actually-enforced' &&
    row.dispatchAdapterExecutionAdapterResultCode !== null &&
    row.dispatchAdapterExecutionAuditEventId !== null &&
    row.dispatchAdapterExecutionRefs.length > 0 &&
    row.adapterDispatchExecutedClaimed
  );
}

function missingExecutionEvidenceIsHonest(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.dispatchAdapterExecutionState ===
      AgentAppGameAdapterDispatchAdapterExecutionState.AdapterExecutionEvidenceMissing &&
    row.dispatchAdapterExecutionDecision ===
      AgentAppGameAdapterDispatchAdapterExecutionDecision.AdapterExecutionEvidenceMissing &&
    row.dispatchAdapterExecutionResultId === null &&
    row.dispatchAdapterExecutionStatus === null &&
    row.dispatchAdapterExecutionAdapterResultCode === null &&
    row.dispatchAdapterExecutionAuditEventId === null &&
    row.dispatchAdapterExecutionRefs.length === 0 &&
    !row.adapterDispatchExecutedClaimed
  );
}

function acceptedDispatchResultMatchesBoundary(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.platform === 'windows' &&
    row.sourceProofEntryId === 'windows-app-game-owned-process-time-limit' &&
    row.dispatchPreflightState === AgentAppGameAdapterDispatchPreflightState.DispatchEligible &&
    row.dispatchDecision === AgentAppGameAdapterDispatchDecision.DispatchEligible &&
    row.dispatchOutcomeState === AgentAppGameAdapterDispatchOutcomeState.DispatchReady &&
    row.dispatchIntentId !== null
  );
}

function acceptedDispatchResultTracksCommand(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.dispatchCommandResultDecision === AgentAppGameAdapterDispatchCommandResultDecision.CommandAccepted &&
    row.enforcementCommandName === AgentEnforcementExecuteCommand &&
    row.enforcementEventName === AgentEnforcementAuditReportedEvent &&
    row.enforcementActionMode === 'terminate-process' &&
    row.dispatchCommandResultId !== null &&
    row.dispatchCommandAuditRefs.length > 0 &&
    row.dispatchCommandTimerRefs.length > 0
  );
}

function acceptedDispatchResultTracksExecutionAudit(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.dispatchExecutionAuditState === AgentAppGameAdapterDispatchExecutionAuditState.ServiceLocalAuditRecorded &&
    row.dispatchExecutionAuditDecision ===
      AgentAppGameAdapterDispatchExecutionAuditDecision.ServiceLocalAuditRecorded &&
    row.dispatchExecutionAuditId !== null &&
    row.dispatchExecutionAuditRefs.length > 0
  );
}

function acceptedDispatchResultKeepsClaimsScoped(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.manualProofRequirements.length === 0 &&
    row.adapterDispatchCommandResultClaimed &&
    row.serviceLocalExecutionAuditClaimed &&
    acceptedExecutionEvidenceIsHonest(row)
  );
}

function blockedDispatchResultTracksCommand(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.dispatchCommandResultDecision === AgentAppGameAdapterDispatchCommandResultDecision.BlockedBeforeCommand &&
    row.enforcementCommandName === null &&
    row.enforcementEventName === null &&
    row.enforcementActionMode === null &&
    row.dispatchCommandResultId === null
  );
}

function blockedDispatchResultTracksExecutionAudit(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.dispatchExecutionAuditState === AgentAppGameAdapterDispatchExecutionAuditState.BlockedBeforeExecutionAudit &&
    row.dispatchExecutionAuditDecision ===
      AgentAppGameAdapterDispatchExecutionAuditDecision.BlockedBeforeExecutionAudit &&
    row.dispatchExecutionAuditId === null &&
    row.dispatchExecutionAuditRefs.length === 0
  );
}

function blockedDispatchResultTracksAdapterExecution(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.dispatchAdapterExecutionState ===
      AgentAppGameAdapterDispatchAdapterExecutionState.BlockedBeforeAdapterExecution &&
    row.dispatchAdapterExecutionDecision ===
      AgentAppGameAdapterDispatchAdapterExecutionDecision.BlockedBeforeAdapterExecution &&
    row.dispatchAdapterExecutionResultId === null &&
    row.dispatchAdapterExecutionStatus === null &&
    row.dispatchAdapterExecutionAdapterResultCode === null &&
    row.dispatchAdapterExecutionAuditEventId === null &&
    row.dispatchAdapterExecutionRefs.length === 0
  );
}

function blockedDispatchResultKeepsClaimsScoped(row: AgentAppGameAdapterDispatchResultRowCandidate): boolean {
  return (
    row.manualProofRequirements.length > 0 &&
    !row.adapterDispatchCommandResultClaimed &&
    !row.adapterDispatchExecutedClaimed &&
    !row.serviceLocalExecutionAuditClaimed
  );
}

File: crates/agent-protocol/tests/contract/app_game_adapter_dispatch_result.rs
use ocentra_eventing::expect_value::ExpectValue;
use ocentra_parent_agent_protocol::AppGameAdapterDispatchResultReadModel;
use ocentra_parent_agent_protocol::AppGameAdapterDispatchResultRow;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_ADAPTER_EXECUTION_DECISION_MISSING;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_ADAPTER_EXECUTION_STATE_MISSING;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_COMMAND_AUDIT_OWNED_PROCESS;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_COMMAND_RESULT_DECISION_ACCEPTED;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_COMMAND_RESULT_STATE_ACCEPTED;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_DECISION_ELIGIBLE;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_DECISION_RECORDED;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_OWNED_PROCESS_ID;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_OWNED_PROCESS_REF;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_STATE_RECORDED;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_INTENT_OWNED_PROCESS_TIME_LIMIT;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_OUTCOME_READY;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_READ_MODEL_ID;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_ROW_ID_PREFIX;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_STATE_ELIGIBLE;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_RESULT_CLAIM_SCOPED_TIMER;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_RESULT_CUSTODY_PREFLIGHT_AND_COMMAND;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_RESULT_FALLBACK_SCOPED_TIMER;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_RESULT_OWNED_PROCESS_ID;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_RESULT_READ_MODEL_ID;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_RESULT_ROW_ID_PREFIX;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_RESULT_STATUS_PARTIAL;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_DISPATCH_TIMER_OWNED_PROCESS;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_PRODUCT_NATIVE_APP;
use ocentra_parent_agent_protocol::APP_GAME_ADAPTER_PRODUCT_NATIVE_GAME;
use ocentra_parent_agent_protocol::APP_GAME_PARENT_PLATFORM_WINDOWS;
use ocentra_parent_agent_protocol::APP_GAME_SCHEMA_VERSION;

const GENERATED_AT: &str = "2026-06-08T10:44:00Z";
const SOURCE_PROOF_ENTRY_ID: &str = "windows-app-game-owned-process-time-limit";
const ADAPTER_CAPABILITY: &str = "app-game-owned-process-time-limit";

#[test]
fn app_game_adapter_dispatch_result_serializes_parent_safe_command_result_rows() {
    let read_model = AppGameAdapterDispatchResultReadModel {
        schema_version: APP_GAME_SCHEMA_VERSION,
        read_model_id: APP_GAME_ADAPTER_DISPATCH_RESULT_READ_MODEL_ID.to_string(),
        generated_at: GENERATED_AT.to_string(),
        source_read_model_ids: vec![
            APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_READ_MODEL_ID.to_string(),
            "agent.enforcement.execute".to_string(),
        ],
        custody_label: APP_GAME_ADAPTER_DISPATCH_RESULT_CUSTODY_PREFLIGHT_AND_COMMAND.to_string(),
        capability_status: APP_GAME_ADAPTER_DISPATCH_RESULT_STATUS_PARTIAL.to_string(),
        returned: 1,
        command_accepted_count: 1,
        blocked_before_command_count: 0,
        execution_audit_recorded_count: 1,
        blocked_before_execution_audit_count: 0,
        adapter_execution_reported_count: 0,
        adapter_execution_evidence_missing_count: 1,
        blocked_before_adapter_execution_count: 0,
        adapter_dispatch_command_result_claimed_count: 1,
        service_local_execution_audit_claimed_count: 1,
        adapter_dispatch_executed_claimed_count: 0,
        broad_installed_app_blocking_claimed: false,
        child_device_delivery_claimed: false,
        platform_enforcement_claimed: false,
        provider_delivery_claimed: false,
        private_diagnostics_claimed: false,
        rows: vec![scoped_command_result_row()],
    };

    let serialized = serde_json::to_value(&read_model).expect_value("dispatch result serializes");

    assert_eq!(
        serialized["readModelId"],
        APP_GAME_ADAPTER_DISPATCH_RESULT_READ_MODEL_ID
    );
    assert_eq!(
        serialized["sourceReadModelIds"][1],
        "agent.enforcement.execute"
    );
    assert_eq!(
        serialized["rows"][0]["dispatchExecutionAuditId"],
        APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_OWNED_PROCESS_ID
    );
    assert_eq!(
        serialized["rows"][0]["adapterDispatchExecutedClaimed"],
        false
    );
}

fn scoped_command_result_row() -> AppGameAdapterDispatchResultRow {
    AppGameAdapterDispatchResultRow {
        schema_version: APP_GAME_SCHEMA_VERSION,
        row_id: format!("{APP_GAME_ADAPTER_DISPATCH_RESULT_ROW_ID_PREFIX}{SOURCE_PROOF_ENTRY_ID}"),
        source_dispatch_preflight_row_id: format!(
            "{APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_ROW_ID_PREFIX}{SOURCE_PROOF_ENTRY_ID}"
        ),
        source_proof_entry_id: SOURCE_PROOF_ENTRY_ID.to_string(),
        platform: APP_GAME_PARENT_PLATFORM_WINDOWS.to_string(),
        product_meanings: vec![
            APP_GAME_ADAPTER_PRODUCT_NATIVE_APP.to_string(),
            APP_GAME_ADAPTER_PRODUCT_NATIVE_GAME.to_string(),
        ],
        adapter_capability: ADAPTER_CAPABILITY.to_string(),
        dispatch_preflight_state: APP_GAME_ADAPTER_DISPATCH_PREFLIGHT_STATE_ELIGIBLE.to_string(),
        dispatch_decision: APP_GAME_ADAPTER_DISPATCH_DECISION_ELIGIBLE.to_string(),
        dispatch_intent_id: Some(
            APP_GAME_ADAPTER_DISPATCH_INTENT_OWNED_PROCESS_TIME_LIMIT.to_string(),
        ),
        dispatch_outcome_state: APP_GAME_ADAPTER_DISPATCH_OUTCOME_READY.to_string(),
        dispatch_command_result_state: APP_GAME_ADAPTER_DISPATCH_COMMAND_RESULT_STATE_ACCEPTED
            .to_string(),
        dispatch_command_result_decision:
            APP_GAME_ADAPTER_DISPATCH_COMMAND_RESULT_DECISION_ACCEPTED.to_string(),
        enforcement_command_name: Some("agent.enforcement.execute".to_string()),
        enforcement_event_name: Some("agent.enforcement.audit.reported".to_string()),
        enforcement_action_mode: Some("terminate-process".to_string()),
        dispatch_command_result_id: Some(
            APP_GAME_ADAPTER_DISPATCH_RESULT_OWNED_PROCESS_ID.to_string(),
        ),
        dispatch_command_audit_refs: vec![
            APP_GAME_ADAPTER_DISPATCH_COMMAND_AUDIT_OWNED_PROCESS.to_string()
        ],
        dispatch_command_timer_refs: vec![APP_GAME_ADAPTER_DISPATCH_TIMER_OWNED_PROCESS.to_string()],
        dispatch_execution_audit_state: APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_STATE_RECORDED
            .to_string(),
        dispatch_execution_audit_decision:
            APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_DECISION_RECORDED.to_string(),
        dispatch_execution_audit_id: Some(
            APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_OWNED_PROCESS_ID.to_string(),
        ),
        dispatch_execution_audit_refs: vec![
            APP_GAME_ADAPTER_DISPATCH_EXECUTION_AUDIT_OWNED_PROCESS_REF.to_string(),
        ],
        dispatch_adapter_execution_state: APP_GAME_ADAPTER_DISPATCH_ADAPTER_EXECUTION_STATE_MISSING
            .to_string(),
        dispatch_adapter_execution_decision:
            APP_GAME_ADAPTER_DISPATCH_ADAPTER_EXECUTION_DECISION_MISSING.to_string(),
        dispatch_adapter_execution_result_id: None,
        dispatch_adapter_execution_status: None,
        dispatch_adapter_execution_adapter_result_code: None,
        dispatch_adapter_execution_audit_event_id: None,
        dispatch_adapter_execution_refs: vec![],
        manual_proof_requirements: vec![],
        claim_boundary: APP_GAME_ADAPTER_DISPATCH_RESULT_CLAIM_SCOPED_TIMER.to_string(),
        fallback_behavior: APP_GAME_ADAPTER_DISPATCH_RESULT_FALLBACK_SCOPED_TIMER.to_string(),
        adapter_dispatch_command_result_claimed: true,
        adapter_dispatch_executed_claimed: false,
        service_local_execution_audit_claimed: true,
        broad_installed_app_blocking_claimed: false,
        child_device_delivery_claimed: false,
        platform_enforcement_claimed: false,
        provider_delivery_claimed: false,
        private_diagnostics_claimed: false,
        last_checked_at: GENERATED_AT.to_string(),
    }
}
